### ISSUE TITLE
[Feature][EPDDisagg] Support a Python API for proxy functionality, including request handling and the ability to launch Encoder and LLM (PD) vLLM instances

### ISSUE THREAD

**Issue by wuhang2014** [hsliuustc0106/vllm#38]
# Design Document: vLLM Disaggregated Proxy System

## Table of Contents
- [Overview](#overview)
- [Architecture](#architecture)
- [System Components](#system-components)
- [Request Flow](#request-flow)
- [Implementation Details](#implementation-details)
- [Performance Characteristics](#performance-characteristics)
- [Use Cases](#use-cases)
- [API Reference](#api-reference)
- [Configuration](#configuration)
- [Error Handling](#error-handling)
- [Future Considerations](#future-considerations)

## Overview

The vLLM Disaggregated Proxy system is a distributed architecture designed to handle Large Language Model (LLM) inference requests across multiple specialized workers. This system implements a proxy pattern that efficiently coordinates between encoding workers, a shared connector for embedding storage, and parallel prefill/decode (PD) workers to handle multimodal AI workloads at scale.

The disaggregated architecture separates multimodal encoding from the main LLM inference, with encode workers computing and storing embeddings in a shared connector, while PD workers operate in parallel and fetch embeddings as needed for multimodal inference.

### Key Benefits
- **Decoupled Storage**: Shared connector enables embedding reuse and persistence
- **Parallel Processing**: Multiple PD workers handle requests concurrently
- **Specialized Encoding**: Dedicated workers for multimodal data preprocessing
- **Embedding Persistence**: Computed embeddings can be cached and reused
- **Horizontal Scalability**: Independent scaling of encoding and PD worker pools
- **Resource Optimization**: Each PD worker handles complete prefill→decode pipeline
- **API Compatibility**: Maintains standard vLLM EngineClient interface

### High-Level Architecture Diagram

```
┌─────────────┐    ┌─────────────┐    ┌─────────────┐
│   Client    │    │   Client    │    │   Client    │
└──────┬──────┘    └──────┬──────┘    └──────┬──────┘
       │                  │                  │
       └──────────────────┼──────────────────┘
                          │
                   ┌──────▼──────┐
                   │    Proxy    │
                   │ (Coordinator)│
                   └──────┬──────┘
                          │
         ┌────────────────┼────────────────┐
         │                │                │
    ┌────▼────┐      ┌────▼────┐      ┌────▼────┐
    │ Encode  │      │ Encode  │      │   PD    │
    │Worker 1 │      │Worker 2 │      │Worker 1 │
    │         │      │         │      │ (P+D)   │
    └────┬────┘      └────┬────┘      └────┬────┘
         │                │                │
         └────────────────┼────────────────┘
                          │
                   ┌──────▼──────┐
                   │  Connector  │
                   │ (Embedding  │
                   │  Storage)   │
                   └─────────────┘

```

### Communication Pattern

```
Client Request → Proxy → [Encode Worker] → Proxy → [PD Worker] → Proxy → Client Response
                    ↓                                 ↓
                            [Connector]
```

For multimodal requests:
1. **Encoding Phase**: Image/multimodal data preprocessing and feature extraction
2. **Prefill Phase**: Process input tokens and build KV cache
3. **Decode Phase**: Iterative token generation with existing KV cache

### 1. Proxy Server

The central coordinator that manages request routing and response aggregation across the disaggregated inference pipeline.

**Core Responsibilities:**
- Request routing and load balancing across worker pools
- Response aggregation and streaming for decode iterations
- Protocol translation between client and specialized workers
- Connection management and error handling across distributed components

**Key Features:**
```python
class Proxy(EngineClient):
    def __init__(self, proxy_addr: str, encode_addr_list: list[str], 
                 pd_addr_list: list[str], model_name: str):
        # ZeroMQ context and socket management for distributed workers
        # Request queue management for async processing
        # Protocol compatibility layer with vLLM EngineClient
```

### 2. Encode Workers

Specialized workers for processing multimodal data before the main inference pipeline.

**Responsibilities:**
- Image preprocessing and feature extraction
- Vision encoder model execution (e.g., CLIP, SigLIP)
- Multimodal data format conversion and encoding
- Feature alignment for language model consumption

**Optimization Focus:**
- GPU memory optimized for vision models
- Batch processing for multiple images
- Efficient serialization of encoded features

### 3. Prefill/Decode (PD) Workers

Workers that handle the core LLM inference pipeline with separated prefill and decode phases.

**Prefill Phase Responsibilities:**
- Process input token sequences
- Build initial Key-Value (KV) cache
- Execute attention computation for input context
- Prepare state for iterative decoding

**Decode Phase Responsibilities:**
- Iterative token generation using existing KV cache
- Sampling and logit processing
- KV cache management and updates
- Response streaming and formatting

**Resource Optimization:**
- High-memory GPUs for large KV caches
- Optimized attention kernels (FlashAttention, PagedAttention)
- Efficient memory management for long sequences

### Prefill/Decode Coordination

**Request Processing Pipeline:**
```python
async def _run_pd(self, request: GenerationRequest, q: asyncio.Queue):
    """
    Coordinate prefill and decode phases:
    1. Send request to PD worker
    2. Worker performs prefill (builds KV cache)
    3. Worker iteratively decodes tokens
    4. Stream responses until completion
    """
    # Route to specific PD worker for session consistency
    idx = (hash(request.request_id) & 0x7FFFFFFF) % len(self.to_pd_sockets)
    socket = self.to_pd_sockets[idx]
    
    # Send generation request
    await socket.send_multipart((RequestType.GENERATION, payload))
    
    # Stream decode responses
    finished = False
    while not finished:
        response = await q.get()
        finished = response.finish_reason is not None
        yield response
```

### Message Protocol

**Protocol Structure optimized for LLM inference:**
```python
# Message format: (RequestType, Payload)
RequestType = Enum('RequestType', ['ENCODE', 'GENERATION'])
ResponseType = Enum('ResponseType', ['GENERATION', 'ENCODE', 'FAILURE'])

# GenerationRequest includes multimodal data and sampling params
class GenerationRequest:
    request_id: str
    prompt: str
    sampling_params: SamplingParams
    multi_modal_data: Optional[dict] = None
```

**Serialization for LLM Data:**
- Uses msgpack for efficient binary serialization
- Optimized for large KV cache state (future enhancement)
- Supports complex multimodal data structures
- Maintains numerical precision for logits and probabilities

### Multimodal Data Handling

**Vision Encoder Integration:**
```python
def _encode_mm_data(mm_data: dict[str, any]) -> dict[str, any]:
    """
    Efficient encoding of multimodal data for LLM consumption:
    - Converts images to format suitable for vision encoders
    - Preserves spatial information and feature dimensions
    - Optimizes for downstream attention computation
    """
    images = mm_data.get("image", [])
    if not isinstance(images, list):
        images = [images]
    
    encoded_images = []
    for img in images:
        if isinstance(img, np.ndarray):
            # Preserve image data for vision encoder processing
            encoded_img = {
                "type": "ndarray",
                "data": img.tobytes(),
                "shape": img.shape,
                "dtype": str(img.dtype),
            }
            encoded_images.append(encoded_img)
    
    return {"image": encoded_images}
```

### Asynchronous Processing for LLM Inference

**Queue-based Request Management:**
```python
class Proxy:
    def __init__(self, ...):
        self.queues: dict[str, asyncio.Queue] = {}
    
    async def generate(self, ...):
        # Create dedicated queue for streaming decode responses
        q = asyncio.Queue()
        self.queues[request_id] = q
        
        # Handle multimodal encoding if needed
        if _has_mm_data(prompt):
            await self._run_encode(request, q)
        
        # Stream prefill and decode responses
        async for pd_response in self._run_pd(request, q):
            yield self._to_request_output(pd_response)
```

### Advantages

1. **Phase-Specific Optimization**
   - Prefill workers optimized for parallel attention computation
   - Decode workers optimized for sequential generation with KV cache
   - Separate resource allocation for compute vs memory-intensive phases

2. **Horizontal Scalability**
   - Independent scaling of encode, prefill, and decode capacity
   - Linear performance improvement with worker count
   - Elastic scaling based on workload characteristics

3. **Resource Specialization**
   - High-memory GPUs for large KV caches in PD workers
   - Vision-optimized GPUs for encode workers
   - CPU workers for preprocessing tasks

4. **Memory Distribution**
   - Distributed KV cache storage across PD workers
   - Reduced memory pressure through specialized allocation
   - Efficient batching within worker types

5. **Throughput Optimization**
   - Parallel prefill processing across workers
   - Concurrent decode iterations
   - Pipeline parallelism for multimodal workflows

### Trade-offs

1. **Network Overhead**
   - Additional serialization for multimodal features
   - Inter-worker communication for large KV states
   - Bandwidth requirements for streaming responses

2. **Complexity**
   - Distributed state management for KV caches
   - Coordination between prefill and decode phases
   - Multiple worker type management

3. **Latency Considerations**
   - Network hops for multimodal encoding
   - Queue management and routing overhead
   - Protocol translation between phases

### Primary Applications

1. **Vision-Language Models (VLMs)**
   - LLaVA, BLIP-2, InstructBLIP deployments
   - Real-time multimodal chat applications
   - Document understanding and analysis

2. **High-Throughput Text Generation**
   - Large-scale API services
   - Batch processing for content generation
   - Enterprise chatbot deployments

3. **Long Context Applications**
   - Document summarization and analysis
   - Code generation and review
   - Research paper processing

4. **Resource-Optimized Deployments**
   - Cost-effective scaling for varying workloads
   - Hybrid cloud-edge deployments
   - GPU resource sharing across teams

### Example Integration Scenarios

**Scenario 1: Multimodal Chat with LLaVA**
```python
# Client code for vision-language interaction
async def chat_with_image(image_path: str, question: str):
    prompt = {
        "prompt": f"USER: <image>\n{question}\nASSISTANT:",
        "multi_modal_data": {
            "image": load_image(image_path)
        }
    }
    
    # Proxy handles: encode → prefill → decode pipeline
    async for response in proxy.generate(
        prompt=prompt,
        sampling_params=SamplingParams(
            max_tokens=256,
            temperature=0.7,
            top_p=0.9
        )
    ):
        print(response.outputs[0].text, end="", flush=True)
```

**Scenario 2: Long Context Document Analysis**
```python
# Efficient processing of long documents
async def analyze_document(document_text: str, questions: list[str]):
    # Prefill phase processes entire document once
    base_prompt = f"Document: {document_text}\n\n"
    
    results = []
    for question in questions:
        full_prompt = base_prompt + f"Question: {question}\nAnswer:"
        
        # KV cache reuse for document context
        async for response in proxy.generate(
            prompt=full_prompt,
            sampling_params=SamplingParams(max_tokens=100),
            request_id=f"doc_analysis_{hash(document_text)}"
        ):
            results.append(response.outputs[0].text)
    
    return results
```

#### `generate()`
```python
async def generate(
    self,
    prompt: PromptType,
    sampling_params: SamplingParams,
    request_id: str,
    lora_request: Optional[LoRARequest] = None,
    trace_headers: Optional[Mapping[str, str]] = None,
    prompt_adapter_request: Optional[PromptAdapterRequest] = None,
    priority: int = 0,
    data_parallel_rank: Optional[int] = None,
) -> AsyncGenerator[RequestOutput, None]:
    """
    Generate text with optional multimodal input through disaggregated pipeline.
    
    Pipeline Flow:
    1. Detect multimodal data and route to encode workers if needed
    2. Route enhanced request to prefill/decode workers
    3. Stream responses from decode phase
    
    Args:
        prompt: Text prompt or dict with multimodal data
        sampling_params: Generation parameters (temperature, top_p, etc.)
        request_id: Unique identifier (enables KV cache reuse)
        
    Yields:
        RequestOutput: Streaming generation results from decode phase
    """
```

#### `shutdown()`
```python
def shutdown(self):
    """
    Clean shutdown of disaggregated proxy resources:
    - Destroys ZMQ context and worker connections
    - Cancels background response handler tasks
    - Removes IPC socket files
    - Ensures proper cleanup of distributed state
    """
```

#### Proxy Initialization for Disaggregated Setup
```python
proxy = Proxy(
    proxy_addr="proxy.ipc",           # IPC address for response aggregation
    encode_addr_list=[                # Vision/multimodal encoding workers
        "encode1.ipc",
        "encode2.ipc"
    ],
    pd_addr_list=[                    # Prefill/decode workers
        "pd1.ipc",
        "pd2.ipc",
        "pd3.ipc"
    ],
    model_name="llava-v1.5-7b-hf"     # Model identifier for compatibility
)
```

### Environment Setup

**Dependencies:**
```bash
pip install pyzmq msgspec numpy torch transformers
# Additional dependencies for multimodal models
pip install Pillow torchvision
```
